### ISSUE TITLE
Make native code super thin?

### ISSUE THREAD

**Issue by fzyzcjy** [bluefireteam/audioplayers#1527]
### Checklist

- [X] I made sure that the issue I am raising doesn't already exist

### Use case / Problem

-

### Proposal / Solution

-

### Example Code

<details>
<summary>Example Code</summary>

```dart
void main() {
}
```

</details>

### Affected platforms

iOS

### Platform details

-

### Working on PR

no way

**Comment by Gustl22** [comment]
Yes the thread stuff currently is a problem in Android. But don't know if it's worth it to use e.g. dart:ffi to interact with the native side. As then probably all modules need a rewrite and some parts are very different on each platform. I also wonder why we use different threads for Android in the first place. This may could be improved in the `exoplayer` rewrite without the need of using ffi.

**Comment by fzyzcjy** [comment]
> But don't know if it's worth it to use e.g. dart:ffi to interact with the native side.

I am also not sure. Btw my proposal above suggests using the current way (or use pigeon if you like - but still the same way).

> As then probably all modules need a rewrite and some parts are very different on each platform

Hopefully a "move" (from native to dart), not a "rewrite" ;)

>  I also wonder why we use different threads for Android in the first place

IMHO it may be caused un-intentionally. 

1. most functions are called by Flutter channels, thus run on thread of flutter channels
2. MediaPlayer callbacks are called on main thread (b/c MediaPlayer cannot find a Looper on the flutter channel threads)

thus there are two threads.

**Comment by Gustl22** [comment]
I also thought about moving the continuous position update logic to the dart thread, but then we would need to request the position every like 200ms (which we also could control from the dart side) and I think it may would be delayed until it reaches the app, in contrast to pushing it from the native side (one call less). But that's something one needs to test.

We also need to define which components exactly can be moved to the dart side without loosing functionality on any platform. Each platform has a different API for their native players which probably can't run unified in dart code. All common elements may are already represented in the [AudioPlayer](https://github.com/bluefireteam/audioplayers/blob/main/packages/audioplayers/lib/src/audioplayer.dart) logic.

**Comment by fzyzcjy** [comment]
> We also need to define which components exactly can be moved to the dart side without loosing functionality on any platform. Each platform has a different API for their native players which probably can't run unified in dart code. All common elements may are already represented in the [AudioPlayer](https://github.com/bluefireteam/audioplayers/blob/main/packages/audioplayers/lib/src/audioplayer.dart?rgh-link-date=2023-05-29T07%3A44%3A53Z) logic.

I guess we can write some different *Dart* code for each platform, just like how we write different *native* code today.

**Comment by Gustl22** [comment]
> * dart starts call -> kotlin receives call: 1ms
> * kotlin ends call -> dart recieves the end of call: 3-4ms

Do you also know the time from "dart starts call to -> kotlin... -> dart receives result", or is this meant by adding these two together?

> I guess we can write some different Dart code for each platform, just like how we write different native code today.

This approach would need to change the general interface logic, as currently we have no dart component in the different platform packages, in contrast to like it's done in [video_player](https://github.com/flutter/plugins/tree/main/packages/video_player/). This also would mean to add much redundant code in each platform implementation on the dart side. But never say no, maybe it's worth it, but honestly I don't think so...

**Comment by fzyzcjy** [comment]
> Do you also know the time from "dart starts call to -> kotlin... -> dart receives result", or is this meant by adding these two together?

What I measured was some long kotlin operations, so I do not report it here since it is unrelated to our scenario here. Yes, the total overhead should add the two IMHO.

> This approach would need to change the general interface logic, as currently we have no dart component in the different platform packages, in contrast to like it's done in [video_player](https://github.com/flutter/plugins/tree/main/packages/video_player/?rgh-link-date=2023-05-29T08%3A18%3A42Z). This also would mean to add much redundant code in each platform implementation on the dart side. But never say no, maybe it's worth it, but honestly I don't think so...

I see. Anyway I may try that approach :) I am using pigeon so the boilerplate seems to be minimal.

**Comment by fzyzcjy** [comment]
To be honest, I implemented what I described above today after posting and discussing here (and it also helps a bit for observations like https://github.com/bluefireteam/audioplayers/issues/1524#issuecomment-1568329028). My implementation seems to be working up to now, so I want to say this may be a viable approach :) If you need any more details, feel free to ping me!

**Comment by Gustl22** [comment]
@fzyzcjy you can post your solution or make a PR to better see what you have done to achive this and if it's applicable for the whole libary, if you're interested :)

**Comment by fzyzcjy** [comment]
Sure! Half a month has passed and I have been moving to work on other parts, but I will try to find a bit of time open-sourcing my code.

The overall idea is very simple: Just do almost nothing on the native side. For example, the `start()` in Android will do nothing but a one-line `exoplayer.start()`.

**Comment by Gustl22** [comment]
I understand the principle. I'm not sure, but I think this needs a separate package / interface to be handled, and due to our capacities I don't think we can contribute much to make it working on every platform. This must be pushed by the community :)

**Comment by jwinarske** [comment]
What about adopting pigeon?  It should be similar to avfoundation video_player.  It might be good to standardize, and it would decrease tech debt down the line.  I have some time mid January.

**Comment by jwinarske** [comment]
I just spent six hours porting your Linux code to ivi-homescreen.  Almost done.  I think pigeon would be a nice improvement.  I can directly use the cpp output from pigeon.  glib support is sitting in a PR.

**Comment by Gustl22** [comment]
@jwinarske I did not use pigeon yet, but it is a valid improvement. If you have the time feel free to open a Draft PR (even if not working), so we can follow the progress. Note that in the approach of video_player, they use a dart layer on the platform implementation, we do not currently have. So we have to validate, how it works out to have "dart" and "native" federated implementations simultaneously for the different platforms.

**Comment by jwinarske** [comment]
@Gustl22 Sure I understand.  I'll put together a WIP.  Might be three weeks from now.

**Comment by jwinarske** [comment]
Linux (glib) PR:  https://github.com/flutter/packages/pull/5100

**Comment by jwinarske** [comment]
My baseline port:  https://github.com/toyota-connected/ivi-homescreen/commit/2c9a70d8737dfc074443ba2ce3bde89319c7be1c
I am seeing general state machine issues around transport state.  More later.
